cocotb/scheduler.py

```diff
@@ -342,114 +344,116 @@ def react(self, trigger, depth=0):
-                    handle.setimmediatevalue(value)
-
-                self._readwrite.unprime()
-
-                return
-
-            # Similarly if we've scheduled our next_timestep on way to readwrite
-            if trigger is self._next_timestep:
-
-                if not self._writes:
-                    self.log.error(
-                        "Moved to next timestep without any pending writes!")
-                else:
-                    self.log.debug(
-                        "Priming ReadWrite trigger so we can playback writes")
-                    self._readwrite.prime(self.react)
-
-                return
+            # work through triggers one by one
+            self._pending_triggers.append(trigger)
+            while self._pending_triggers:
+                trigger = self._pending_triggers.pop(0)
 
-            if trigger not in self._trigger2coros:
-
-                # GPI triggers should only be ever pending if there is an
-                # associated coroutine waiting on that trigger, otherwise it would
-                # have been unprimed already
-                if isinstance(trigger, GPITrigger):
-                    self.log.critical(
-                        "No coroutines waiting on trigger that fired: %s" %
-                        str(trigger))
-
-                    trigger.log.info("I'm the culprit")
-                # For Python triggers this isn't actually an error - we might do
-                # event.set() without knowing whether any coroutines are actually
-                # waiting on this event, for example
-                elif _debug:
-                    self.log.debug(
-                        "No coroutines waiting on trigger that fired: %s" %
-                        str(trigger))
+                if trigger is self._readonly:
+                    self._mode = Scheduler._MODE_READONLY
+                # Only GPI triggers affect the simulator scheduling mode
+                elif isinstance(trigger, GPITrigger):
+                    self._mode = Scheduler._MODE_NORMAL
 
-                return
+                # We're the only source of ReadWrite triggers which are only used for
+                # playing back any cached signal updates
+                if trigger is self._readwrite:
 
-            # Scheduled coroutines may append to our waiting list so the first
-            # thing to do is pop all entries waiting on this trigger.
-            scheduling = self._trigger2coros.pop(trigger)
+                    if _debug:
+                        self.log.debug("Writing cached signal updates")
+
+                    while self._writes:
+                        handle, value = self._writes.popitem()
+                        handle.setimmediatevalue(value)
+
+                    self._readwrite.unprime()
+
+                    # this should be the only trigger, probably?
+                    assert not self._pending_triggers
+                    return
```

Comment:
These assertions are only here for the sake of proving my hunch in CI. If they pass then I'll attempt a further refinement.